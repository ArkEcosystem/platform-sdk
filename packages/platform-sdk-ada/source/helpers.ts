import { Coins } from "@arkecosystem/platform-sdk";
import CardanoWasm, {
	Address,
	BigNum,
	Bip32PrivateKey,
	Bip32PublicKey,
	TransactionBuilder,
	TransactionInput,
} from "@emurgo/cardano-serialization-lib-nodejs";
import { Buffer } from "buffer";

import { HttpClient } from "@arkecosystem/platform-sdk-http";
import { fetchUsedAddressesData } from "./graphql-helpers";
import { addressFromAccountExtPublicKey, deriveAddress, deriveChangeKey, deriveSpendKey } from "./shelley";
import { createValue } from "./transaction.helpers";
import { UnspentTransaction } from "./transaction.models";

export const usedAddressesForAccount = async (
	config: Coins.ConfigRepository,
	httpClient: HttpClient,
	accountPublicKey: string,
) => {
	const networkId: string = config.get<string>("network.meta.networkId");
	const usedSpendAddresses: Set<string> = new Set<string>();
	const usedChangeAddresses: Set<string> = new Set<string>();

	let offset = 0;
	let exhausted = false;
	do {
		const spendAddresses: string[] = await addressesChunk(networkId, accountPublicKey, false, offset);
		const changeAddresses: string[] = await addressesChunk(networkId, accountPublicKey, true, offset);

		const allAddresses = spendAddresses.concat(changeAddresses);
		const usedAddresses: string[] = await fetchUsedAddressesData(config, httpClient, allAddresses);

		spendAddresses
			.filter((sa) => usedAddresses.find((ua) => ua === sa) !== undefined)
			.forEach((sa) => usedSpendAddresses.add(sa));
		changeAddresses
			.filter((sa) => usedAddresses.find((ua) => ua === sa) !== undefined)
			.forEach((sa) => usedChangeAddresses.add(sa));

		exhausted = usedAddresses.length === 0;
		offset += 20;
	} while (!exhausted);
	return { usedSpendAddresses, usedChangeAddresses };
};

export const addressesChunk = async (
	networkId: string,
	accountPublicKey: string,
	isChange: boolean,
	offset: number,
): Promise<string[]> => {
	const publicKey = Buffer.from(accountPublicKey, "hex");

	const addresses: string[] = [];
	for (let i = offset; i < offset + 20; ++i) {
		addresses.push(addressFromAccountExtPublicKey(publicKey, isChange, i, networkId));
	}
	return addresses;
};

export const addUtxoInput = (
	txBuilder: TransactionBuilder,
	input: UnspentTransaction,
): { added: boolean; amount: BigNum; fee: BigNum } => {
	const wasmAddr = Address.from_bech32(input.address);
	const txInput = utxoToTxInput(input);
	const wasmAmount = createValue(input.value);

	// ignore UTXO that contribute less than their fee if they also don't contribute a token
	const feeForInput = txBuilder.fee_for_input(wasmAddr, txInput, wasmAmount);

	const skipped = feeForInput.compare(BigNum.from_str(input.value)) > 0;
	if (!skipped) {
		txBuilder.add_input(wasmAddr, txInput, wasmAmount);
	}
	return { added: !skipped, amount: BigNum.from_str(input.value), fee: feeForInput };
};

export const utxoToTxInput = (utxo: UnspentTransaction): TransactionInput => {
	return CardanoWasm.TransactionInput.new(
		CardanoWasm.TransactionHash.from_bytes(Buffer.from(utxo.txHash, "hex")),
		parseInt(utxo.index),
	);
};

export const deriveAddressesAndSigningKeys = async (
	publicKey: Bip32PublicKey,
	networkId,
	accountKey: Bip32PrivateKey,
) => {
	const addresses: { [index: number]: {} } = { 0: {}, 1: {} };
	for (let i = 0; i < 20; ++i) {
		addresses[0][deriveAddress(publicKey, false, i, networkId)] = deriveSpendKey(accountKey, i);
		addresses[1][deriveAddress(publicKey, true, i, networkId)] = deriveChangeKey(accountKey, i);
	}
	return addresses;
};
