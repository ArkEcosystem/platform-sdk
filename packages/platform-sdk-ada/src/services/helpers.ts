import { Coins, Contracts } from "@arkecosystem/platform-sdk";
import { Arr } from "@arkecosystem/platform-sdk-support";

export const postGraphql = async (config: Coins.Config, query: string): Promise<Record<string, any>> => {
	console.log('query', query);
	const keyValuePair = await config
		.get<Contracts.HttpClient>("httpClient")
		.post(Arr.randomElement(config.get<string[]>("network.networking.hostsArchival")), { query });
	console.log('keyValuePair', JSON.stringify(keyValuePair.json(), null, 2));
	return keyValuePair.json().data;
};
