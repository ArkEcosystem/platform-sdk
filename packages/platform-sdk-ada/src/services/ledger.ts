import { Coins, Services } from "@arkecosystem/platform-sdk";

export class LedgerService extends Services.AbstractLedgerService {
	public static async __construct(config: Coins.ConfigRepository): Promise<LedgerService> {
		return new LedgerService();
	}
}
