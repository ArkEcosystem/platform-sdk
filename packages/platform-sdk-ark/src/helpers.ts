import { Coins } from "@arkecosystem/platform-sdk";
import { Arr } from "@arkecosystem/platform-sdk-support";

export const getPeerFromConfig = (config: Coins.Config): string => {
	try {
		return config.get<string>("peer");
	} catch {
		return `${Arr.randomElement(config.get<string[]>("network.networking.hosts"))}/api`;
	}
}
