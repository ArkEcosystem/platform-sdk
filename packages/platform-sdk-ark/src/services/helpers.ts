import { Managers } from "@arkecosystem/crypto";
import { Coins, Contracts } from "@arkecosystem/platform-sdk";
import { Arr } from "@arkecosystem/platform-sdk-support";

export const retrieveCryptoConfiguration = async (config: Coins.Config): Promise<{ crypto; peer; status }> => {
	const http: Contracts.HttpClient = config.get<Contracts.HttpClient>("httpClient");

	let peer: string;
	try {
		peer = config.get<string>("peer");
	} catch {
		peer = `${Arr.randomElement(config.get<string[]>("network.networking.hosts"))}/api`;
	}

	const crypto: any = (await http.get(`${peer}/node/configuration/crypto`)).json();
	Managers.configManager.setConfig(crypto.data);

	const status: any = (await http.get(`${peer}/node/syncing`)).json();
	Managers.configManager.setHeight(status.data.height);

	return { crypto, peer, status };
};

export const applyCryptoConfiguration = ({ crypto, status }): void => {
	Managers.configManager.setConfig(crypto.data);
	Managers.configManager.setHeight(status.data.height);
};
