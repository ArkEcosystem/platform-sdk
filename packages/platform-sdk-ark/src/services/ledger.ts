import { ARKTransport } from "@arkecosystem/ledger-transport";
import { Coins, Contracts } from "@arkecosystem/platform-sdk";
import LedgerTransport from "@ledgerhq/hw-transport-node-hid-singleton";

export class LedgerService implements Contracts.LedgerService {
	#ledger: LedgerTransport;
	#transport!: ARKTransport;

	private constructor(transport: Contracts.LedgerTransport) {
		this.#ledger = transport;
	}

	public static async construct(config: Coins.Config): Promise<LedgerService> {
		try {
			return new LedgerService(config.get("services.ledger.transport"));
		} catch {
			return new LedgerService(LedgerTransport);
		}
	}

	public async destruct(): Promise<void> {
		await this.disconnect();
	}

	public async connect(): Promise<void> {
		this.#ledger = await this.#ledger.open();
		this.#transport = new ARKTransport(this.#ledger);
	}

	public async disconnect(): Promise<void> {
		await this.#ledger.close();
	}

	public async getVersion(): Promise<string> {
		return this.#transport.getVersion();
	}

	public async getPublicKey(path: string): Promise<string> {
		return this.#transport.getPublicKey(path);
	}

	public async signTransaction(path: string, payload: Buffer): Promise<string> {
		return this.#transport.signTransaction(path, payload);
	}

	public async signTransactionWithSchnorr(path: string, payload: Buffer): Promise<string> {
		return this.#transport.signTransaction(path, payload);
	}

	public async signMessage(path: string, payload: Buffer): Promise<string> {
		return this.#transport.signMessage(path, payload);
	}

	public async signMessageWithSchnorr(path: string, payload: Buffer): Promise<string> {
		return this.#transport.signMessage(path, payload);
	}
}
