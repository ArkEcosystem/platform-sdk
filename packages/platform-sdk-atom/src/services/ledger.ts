import { Coins, Contracts, Exceptions } from "@arkecosystem/platform-sdk";
import { BIP44 } from "@arkecosystem/platform-sdk-crypto";
import Cosmos from "ledger-cosmos-js";

export class LedgerService implements Contracts.LedgerService {
	#ledger: Contracts.LedgerTransport;
	#transport!: Cosmos;

	public static async construct(config: Coins.Config): Promise<LedgerService> {
		return new LedgerService();
	}

	public async destruct(): Promise<void> {
		await this.disconnect();
	}

	public async connect(transport: Contracts.LedgerTransport): Promise<void> {
		this.#ledger = await transport.create();
		this.#transport = new Cosmos(this.#ledger);
	}

	public async disconnect(): Promise<void> {
		await this.#ledger.close();
	}

	public async getVersion(): Promise<string> {
		const res = await this.#transport.getVersion();

		return `${res.major}.${res.minor}.${res.patch}`;
	}

	public async getPublicKey(path: string): Promise<string> {
		const pathArray: number[] = Object.values(BIP44.parse(path));
		const { compressed_pk } = await this.#transport.publicKey(pathArray);

		return compressed_pk.toString("hex");
	}

	public async signTransaction(path: string, payload: Buffer): Promise<string> {
		const pathArray: number[] = Object.values(BIP44.parse(path));
		const { signature } = await this.#transport.sign(pathArray, payload.toString());

		return signature.toString("hex");
	}

	public async signTransactionWithSchnorr(path: string, payload: Buffer): Promise<string> {
		throw new Exceptions.NotImplemented(this.constructor.name, "signTransactionWithSchnorr");
	}

	public async signMessage(path: string, payload: Buffer): Promise<string> {
		throw new Exceptions.NotImplemented(this.constructor.name, "signMessage");
	}

	public async signMessageWithSchnorr(path: string, payload: Buffer): Promise<string> {
		throw new Exceptions.NotImplemented(this.constructor.name, "signMessageWithSchnorr");
	}
}
