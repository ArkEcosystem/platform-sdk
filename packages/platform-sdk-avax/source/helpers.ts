import { Coins, Helpers, Services } from "@arkecosystem/platform-sdk";
import { BIP39, BIP44 } from "@arkecosystem/platform-sdk-crypto";
import { Avalanche, BinTools, Buffer } from "avalanche";
import { AVMAPI, KeyPair } from "avalanche/dist/apis/avm";
import { InfoAPI } from "avalanche/dist/apis/info";
import { PlatformVMAPI } from "avalanche/dist/apis/platformvm";
import HDKey from "hdkey";
import urlParseLax from "url-parse-lax";

export const useAvalanche = (config: Coins.ConfigRepository): Avalanche => {
	const { hostname: host, port, protocol } = urlParseLax(Helpers.randomHostFromConfig(config));

	return new Avalanche(
		host,
		port,
		protocol.replace(":", ""),
		parseInt(config.get("network.meta.networkId")),
		config.get("network.meta.blockchainId"),
	);
};

export const useInfo = (config: Coins.ConfigRepository): InfoAPI => useAvalanche(config).Info();

export const useXChain = (config: Coins.ConfigRepository): AVMAPI => useAvalanche(config).XChain();

export const usePChain = (config: Coins.ConfigRepository): PlatformVMAPI => useAvalanche(config).PChain();

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useKeychain = (config: Coins.ConfigRepository) => useXChain(config).keyChain();

export const cb58Decode = (value: string): Buffer => BinTools.getInstance().cb58Decode(value);

export const cb58Encode = (value: Buffer): string => BinTools.getInstance().cb58Encode(value);

// Crypto
export const keyPairFromMnemonic = (
	config: Coins.ConfigRepository,
	mnemonic: string,
	options?: Services.IdentityOptions,
): { child: KeyPair; path: string } => {
	const path = BIP44.stringify({
		coinType: config.get(Coins.ConfigKey.Slip44),
		account: options?.bip44?.account,
		index: options?.bip44?.addressIndex,
	});

	return {
		child: useKeychain(config).importKey(
			// @ts-ignore
			HDKey.fromMasterSeed(BIP39.toSeed(mnemonic)).derive(path).privateKey,
		),
		path,
	};
};
