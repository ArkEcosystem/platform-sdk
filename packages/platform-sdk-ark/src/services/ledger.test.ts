import "jest-extended";
import { createTransportReplayer, RecordStore } from "@ledgerhq/hw-transport-mocker";
import { LedgerService } from "./ledger";
import { ledger } from "../../test/fixtures/ledger";
import { createConfig } from "../../test/helpers";

const createMockService = async (record: string) => {
	const transport = await LedgerService.construct(createConfig());

	await transport.connect(createTransportReplayer(RecordStore.fromString(record)));

	return transport;
};

describe("constructor", () => {
	it("should pass with an empty configuration", async () => {
		const transport = await LedgerService.construct(
			createConfig({
				services: {
					ledger: {},
				},
			}),
		);

		expect(transport).toBeInstanceOf(LedgerService);
	});
});

describe("destruct", () => {
	it("should pass with a resolved transport closure", async () => {
		const ark = await createMockService("");

		await expect(ark.destruct()).resolves.toBeUndefined();
	});
});

describe("getVersion", () => {
	it("should pass with an app version", async () => {
		const ark = await createMockService(ledger.appVersion.record);

		await expect(ark.getVersion()).resolves.toEqual(ledger.appVersion.result);
	});
});

describe("getPublicKey", () => {
	it("should pass with a compressed publicKey", async () => {
		const ark = await createMockService(ledger.publicKey.record);

		await expect(ark.getPublicKey(ledger.bip44.path)).resolves.toEqual(ledger.publicKey.result);
	});
});

describe("signTransaction", () => {
	it("should pass with an ecdsa signature", async () => {
		const ark = await createMockService(ledger.transaction.ecdsa.record);

		await expect(
			ark.signTransaction(ledger.bip44.path, Buffer.from(ledger.transaction.ecdsa.payload, "hex")),
		).resolves.toEqual(ledger.transaction.ecdsa.result);
	});
});

describe("signTransactionWithSchnorr", () => {
	it("should pass with a schnorr signature", async () => {
		const ark = await createMockService(ledger.transaction.schnorr.record);

		await expect(
			ark.signTransactionWithSchnorr(ledger.bip44.path, Buffer.from(ledger.transaction.schnorr.payload, "hex")),
		).resolves.toEqual(ledger.transaction.schnorr.result);
	});
});

describe("signMessage", () => {
	it("should pass with an ecdsa signature", async () => {
		const ark = await createMockService(ledger.message.ecdsa.record);

		await expect(
			ark.signMessage(ledger.bip44.path, Buffer.from(ledger.message.ecdsa.payload, "hex")),
		).resolves.toEqual(ledger.message.ecdsa.result);
	});
});

describe("signMessageWithSchnorr", () => {
	it("should pass with a schnorr signature", async () => {
		const ark = await createMockService(ledger.message.schnorr.record);

		await expect(
			ark.signMessageWithSchnorr(ledger.bip44.path, Buffer.from(ledger.message.schnorr.payload, "hex")),
		).resolves.toEqual(ledger.message.schnorr.result);
	});
});
