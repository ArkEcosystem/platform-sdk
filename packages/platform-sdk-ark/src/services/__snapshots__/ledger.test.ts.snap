// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scan should return scanned wallet 1`] = `
Object {
  "44'/1'/0'/0/0": WalletData {
    "data": Object {
      "address": "D9xJncW4ECUSJQWeLP7wncxhDTvNeg2HNK",
      "attributes": Object {},
      "balance": "15000000000",
      "nonce": "0",
    },
  },
  "44'/1'/1'/0/0": WalletData {
    "data": Object {
      "address": "DFgggtreMXQNQKnxHddvkaPHcQbRdK3jyJ",
      "balance": 0,
      "publicKey": "0348a63f3d9c0050575cbfe70de1722fd73ba205a9d2f7150d0725b99abb331892",
    },
  },
}
`;

exports[`scan should return scanned wallet 2`] = `
Object {
  "address": "DFgggtreMXQNQKnxHddvkaPHcQbRdK3jyJ",
  "balance": BigNumber {},
  "nonce": BigNumber {},
  "publicKey": "0348a63f3d9c0050575cbfe70de1722fd73ba205a9d2f7150d0725b99abb331892",
  "rank": undefined,
  "username": undefined,
  "votes": undefined,
}
`;

exports[`scan should return scanned wallet 3`] = `
Object {
  "address": "D9xJncW4ECUSJQWeLP7wncxhDTvNeg2HNK",
  "balance": BigNumber {},
  "nonce": BigNumber {},
  "publicKey": undefined,
  "rank": undefined,
  "username": undefined,
  "votes": undefined,
}
`;
